File: SPOJ/Graph/AKBAR.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cmath>
#include <cstdio>

using namespace std;

int main(int argc, char const *argv[]) {
	int t;
	scanf("%d", &t);
	while(t--){

		long long n,r,m;
		scanf("%lld", &n); scanf("%lld", &r); scanf("%lld", &m);

		vector<long long> g[n+1];
		vector<bool> vis1(n+1, false);
		vector<long long> treno(n+1,1000000);
	
		long long a,b;
		for(long long i=0;i<r; ++i){
			scanf("%lld", &a); scanf("%lld", &b);
			g[a].push_back(b);
			g[b].push_back(a);
		}
		
		int ans = 1;
		long long nono = 0;
		for(int i=0; i<m; ++i){
			scanf("%lld", &a); scanf("%lld", &b);

			if(ans==0) continue;

			if(treno[a]<nono){
				ans = 0;
				continue;
			}

			vector<bool> vis2(n+1, false);
			vector<long long> le(n+1,-1);

			queue<long long> data;
		
			vis2[a] = true;
			vis1[a] = true;
			le[a] = 0;
			data.push(a);
			
			while(!data.empty()){
				long long node = data.front();
				data.pop();

				for(vector<long long>::iterator i = g[node].begin(); i!=g[node].end() && le[node]<b; ++i) {

					if(treno[*i]<nono){
						ans = 0;
						break;
					}
				
					if(!vis2[*i]){
						vis1[*i] = vis2[*i] = true;
						treno[*i] = nono; 
						le[*i] = le[node]+1;
						data.push(*i);
					}	
				}
				if(ans == 0) break;
			}
			nono++;
		}

		if(ans) {
			for(int i=1; i<=n; ++i){
				if(!vis1[i]){
					ans = 0;
					break;
				}
			}

		}
		if(ans) printf("YES\n");
		else printf("NO\n");
	}

	return 0;
}

File: SPOJ/Graph/AKBAR.py
from Queue import Queue
ri=raw_input

T = int(ri())
for t in range(T):
	n,r,m = map(int,ri().split())
	
	g = [ [] for _ in range(n+1)]
	vis1 = [ 0 for _ in range(n+1)]

	for i in range(r):
		a,b = map(int,ri().split())
		g[a].append(b)
		g[b].append(a)
	
	for j in range(m):
		a,b = map(int,ri().split())

		vis2 = [ False for _ in range(n+1)]
		level = [ -1 for _ in range(n+1)]

		q1 = Queue()
		q2 = Queue()

		vis2[a] = True
		level[a] = 0;

		q1.put(a)
		q2.put(a)

		while(not q1.empty()):
			node = q1.get()

			if level[node] < b:
				continue

			for i in g[node]:
				if not vis2[i]:
					vis2[i] = True
					level[i] = level[node] + 1
					q1.put(i)
					q2.put(i)

		while(not q2.empty()):
			vis1[q2.get()]+=1

	ans = True

	for i in range(1,n):
		if vis1[i]==0 or vis1[i]>1:
			ans = False
			break

	if ans : 
		print "Yes"
	else:
		print "No"